### PR TITLE
[SLP]Make PHICompare comparator follow weak strict ordering requirement

### PR DESCRIPTION
**Pull request by alexey-bataev**

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -5454,21 +5470,60 @@ BoUpSLP::getReorderingData(const TreeEntry &TE, bool TopToBottom) {
...
       auto *FirstUserOfPhi1 = cast<Instruction>(*V1->user_begin());
       auto *FirstUserOfPhi2 = cast<Instruction>(*V2->user_begin());
-      if (auto *IE1 = dyn_cast<InsertElementInst>(FirstUserOfPhi1))
-        if (auto *IE2 = dyn_cast<InsertElementInst>(FirstUserOfPhi2)) {
-          if (!areTwoInsertFromSameBuildVector(
-                  IE1, IE2,
-                  [](InsertElementInst *II) { return II->getOperand(0); }))
-            return I1 < I2;
+      if (FirstUserOfPhi1->getParent() != FirstUserOfPhi2->getParent())
+        return DT->dominates(FirstUserOfPhi1->getParent(),
```
Comment (efriedma-quic) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:5474]: dominates() is not a strict weak ordering.  Use getDFSNumIn() or something like that.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -5454,21 +5470,60 @@ BoUpSLP::getReorderingData(const TreeEntry &TE, bool TopToBottom) {
...
+        auto *P1 = dyn_cast<Argument>(EE1->getOperand(0));
+        auto *P2 = dyn_cast<Argument>(EE2->getOperand(0));
+        if (P1 && !P2)
+          return true;
+        if (!P1 && P2)
+          return false;
+        if (P1 && P2)
+          return P1->getArgNo() < P2->getArgNo();
+        // TODO: add analysis for other value kinds.
+        return EE1->getOperand(0)->getValueID() <
```
Comment (efriedma-quic) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:5523]: Maybe simpler to just say all extractelements where the first operand is not an instruction/argument compare equal to each other, so we don't need to come up with an ordering.  It shouldn't really come up, anyway; anything else has to be a Constant.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -5454,21 +5481,58 @@ BoUpSLP::getReorderingData(const TreeEntry &TE, bool TopToBottom) {
...
+        auto *P2 = dyn_cast<Argument>(EE2->getOperand(0));
+        if (P1 && !P2)
+          return true;
+        if (!P1 && P2)
+          return false;
+        if (P1 && P2)
+          return P1->getArgNo() < P2->getArgNo();
+        return false;
+      }
+      return false;
```
Comment (efriedma-quic) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:5535]: This doesn't quite work... if the first operand isn't an instruction or an argument, we have to always return false. So this needs to happen before the `if (EE1->getOperand(0) == EE2->getOperand(0))`.

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:5535]: I think it already does this. if (EE1->getOperand(0) == EE2->getOperand(0)) occurs only if both users are extractelements, argument comparison occurs only if both users are arguments. If the first is extract/argument but the second is not - return true. If the first is not extract/argument - return false. Am I missing something?

Comment (efriedma-quic) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:5534]: Suppose you have three extractelements with a constant as the first operand.  Two of them have A as the first operand, the other one has B as the first operand.  They all need to compare equal, even if the extract indexes are different.

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:5534]: Ah, missed it, fixed

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Comment by mikaelholmen** [comment]
Hi,

I bisected a crash back to this commit.
Reproduce with:
```opt -passes="slp-vectorizer" -mtriple=aarch64 bbi-104215.ll -o /dev/null```
Result:
```
opt: ../lib/Transforms/Vectorize/SLPVectorizer.cpp:5730: auto llvm::slpvectorizer::BoUpSLP::getReorderingData(const TreeEntry &, bool)::(anonymous class)::operator()(BasicBlock *, BasicBlock *) const: Assertion `NodeB && "Should only process reachable instructions"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: build-all/bin/opt -passes=slp-vectorizer -mtriple=aarch64 bbi-104215.ll -o /dev/null
1.	Running pass "function(slp-vectorizer)" on module "bbi-104215.ll"
2.	Running pass "slp-vectorizer" on function "f_180_2259"
 #0 0x00005602853a5026 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (build-all/bin/opt+0x4646026)
 #1 0x00005602853a2a6e llvm::sys::RunSignalHandlers() (build-all/bin/opt+0x4643a6e)
 #2 0x00005602853a58a9 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
 #3 0x00007f5db37acd10 __restore_rt (/lib64/libpthread.so.0+0x12d10)
 #4 0x00007f5db114c52f raise (/lib64/libc.so.6+0x4e52f)
 #5 0x00007f5db111fe65 abort (/lib64/libc.so.6+0x21e65)
 #6 0x00007f5db111fd39 _nl_load_domain.cold.0 (/lib64/libc.so.6+0x21d39)
 #7 0x00007f5db1144e86 (/lib64/libc.so.6+0x46e86)
 #8 0x0000560286e891cd llvm::slpvectorizer::BoUpSLP::getReorderingData(llvm::slpvectorizer::BoUpSLP::TreeEntry const&, bool)::$_6::operator()(llvm::BasicBlock*, llvm::BasicBlock*) const SLPVectorizer.cpp:0:0
 #9 0x0000560286e8887d void std::__merge_adaptive<unsigned int*, long, unsigned int*, __gnu_cxx::__ops::_Iter_comp_iter<llvm::slpvectorizer::BoUpSLP::getReorderingData(llvm::slpvectorizer::BoUpSLP::TreeEntry const&, bool)::$_7>>(unsigned int*, unsigned int*, unsigned int*, long, long, unsigned int*, __gnu_cxx::__ops::_Iter_comp_iter<llvm::slpvectorizer::BoUpSLP::getReorderingData(llvm::slpvectorizer::BoUpSLP::TreeEntry const&, bool)::$_7>) SLPVectorizer.cpp:0:0
#10 0x0000560286dd42b3 llvm::slpvectorizer::BoUpSLP::getReorderingData(llvm::slpvectorizer::BoUpSLP::TreeEntry const&, bool) (build-all/bin/opt+0x60752b3)
#11 0x0000560286dd8354 llvm::slpvectorizer::BoUpSLP::reorderTopToBottom() (build-all/bin/opt+0x6079354)
#12 0x0000560286e3ff0c llvm::SLPVectorizerPass::tryToVectorizeList(llvm::ArrayRef<llvm::Value*>, llvm::slpvectorizer::BoUpSLP&, bool) (build-all/bin/opt+0x60e0f0c)
#13 0x0000560286e45a04 bool tryToVectorizeSequence<llvm::Value>(llvm::SmallVectorImpl<llvm::Value*>&, llvm::function_ref<bool (llvm::Value*, llvm::Value*)>, llvm::function_ref<bool (llvm::Value*, llvm::Value*)>, llvm::function_ref<bool (llvm::ArrayRef<llvm::Value*>, bool)>, bool, llvm::slpvectorizer::BoUpSLP&) SLPVectorizer.cpp:0:0
#14 0x0000560286e3819e llvm::SLPVectorizerPass::vectorizeChainsInBlock(llvm::BasicBlock*, llvm::slpvectorizer::BoUpSLP&) (build-all/bin/opt+0x60d919e)
#15 0x0000560286e35c64 llvm::SLPVectorizerPass::runImpl(llvm::Function&, llvm::ScalarEvolution*, llvm::TargetTransformInfo*, llvm::TargetLibraryInfo*, llvm::AAResults*, llvm::LoopInfo*, llvm::DominatorTree*, llvm::AssumptionCache*, llvm::DemandedBits*, llvm::OptimizationRemarkEmitter*) (build-all/bin/opt+0x60d6c64)
#16 0x0000560286e351e7 llvm::SLPVectorizerPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x60d61e7)
#17 0x000056028682b6cd llvm::detail::PassModel<llvm::Function, llvm::SLPVectorizerPass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilderPipelines.cpp:0:0
#18 0x00005602855c75c7 llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x48685c7)
#19 0x00005602868220fd llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilderPipelines.cpp:0:0
#20 0x00005602855cc19e llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x486d19e)
#21 0x00005602868187ed llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) PassBuilderPipelines.cpp:0:0
#22 0x00005602855c62b7 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x48672b7)
#23 0x00005602867a3b5c llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (build-all/bin/opt+0x5a44b5c)
#24 0x0000560285367c12 optMain (build-all/bin/opt+0x4608c12)
#25 0x00007f5db11387e5 __libc_start_main (/lib64/libc.so.6+0x3a7e5)
#26 0x000056028536582e _start (build-all/bin/opt+0x460682e)
Abort (core dumped)
```
[bbi-104215.ll.gz](https://github.com/user-attachments/files/18839931/bbi-104215.ll.gz)

**Comment by mikaelholmen** [comment]
> > Hi,
> > I bisected a crash back to this commit. Reproduce with: `opt -passes="slp-vectorizer" -mtriple=aarch64 bbi-104215.ll -o /dev/null` Result:
> > ```
> > opt: ../lib/Transforms/Vectorize/SLPVectorizer.cpp:5730: auto llvm::slpvectorizer::BoUpSLP::getReorderingData(const TreeEntry &, bool)::(anonymous class)::operator()(BasicBlock *, BasicBlock *) const: Assertion `NodeB && "Should only process reachable instructions"' failed.
> > PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
> > Stack dump:
> > 0.	Program arguments: build-all/bin/opt -passes=slp-vectorizer -mtriple=aarch64 bbi-104215.ll -o /dev/null
> > 1.	Running pass "function(slp-vectorizer)" on module "bbi-104215.ll"
> > 2.	Running pass "slp-vectorizer" on function "f_180_2259"
> >  #0 0x00005602853a5026 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (build-all/bin/opt+0x4646026)
> >  #1 0x00005602853a2a6e llvm::sys::RunSignalHandlers() (build-all/bin/opt+0x4643a6e)
> >  #2 0x00005602853a58a9 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
> >  #3 0x00007f5db37acd10 __restore_rt (/lib64/libpthread.so.0+0x12d10)
> >  #4 0x00007f5db114c52f raise (/lib64/libc.so.6+0x4e52f)
> >  #5 0x00007f5db111fe65 abort (/lib64/libc.so.6+0x21e65)
> >  #6 0x00007f5db111fd39 _nl_load_domain.cold.0 (/lib64/libc.so.6+0x21d39)
> >  #7 0x00007f5db1144e86 (/lib64/libc.so.6+0x46e86)
> >  #8 0x0000560286e891cd llvm::slpvectorizer::BoUpSLP::getReorderingData(llvm::slpvectorizer::BoUpSLP::TreeEntry const&, bool)::$_6::operator()(llvm::BasicBlock*, llvm::BasicBlock*) const SLPVectorizer.cpp:0:0
> >  #9 0x0000560286e8887d void std::__merge_adaptive<unsigned int*, long, unsigned int*, __gnu_cxx::__ops::_Iter_comp_iter<llvm::slpvectorizer::BoUpSLP::getReorderingData(llvm::slpvectorizer::BoUpSLP::TreeEntry const&, bool)::$_7>>(unsigned int*, unsigned int*, unsigned int*, long, long, unsigned int*, __gnu_cxx::__ops::_Iter_comp_iter<llvm::slpvectorizer::BoUpSLP::getReorderingData(llvm::slpvectorizer::BoUpSLP::TreeEntry const&, bool)::$_7>) SLPVectorizer.cpp:0:0
> > #10 0x0000560286dd42b3 llvm::slpvectorizer::BoUpSLP::getReorderingData(llvm::slpvectorizer::BoUpSLP::TreeEntry const&, bool) (build-all/bin/opt+0x60752b3)
> > #11 0x0000560286dd8354 llvm::slpvectorizer::BoUpSLP::reorderTopToBottom() (build-all/bin/opt+0x6079354)
> > #12 0x0000560286e3ff0c llvm::SLPVectorizerPass::tryToVectorizeList(llvm::ArrayRef<llvm::Value*>, llvm::slpvectorizer::BoUpSLP&, bool) (build-all/bin/opt+0x60e0f0c)
> > #13 0x0000560286e45a04 bool tryToVectorizeSequence<llvm::Value>(llvm::SmallVectorImpl<llvm::Value*>&, llvm::function_ref<bool (llvm::Value*, llvm::Value*)>, llvm::function_ref<bool (llvm::Value*, llvm::Value*)>, llvm::function_ref<bool (llvm::ArrayRef<llvm::Value*>, bool)>, bool, llvm::slpvectorizer::BoUpSLP&) SLPVectorizer.cpp:0:0
> > #14 0x0000560286e3819e llvm::SLPVectorizerPass::vectorizeChainsInBlock(llvm::BasicBlock*, llvm::slpvectorizer::BoUpSLP&) (build-all/bin/opt+0x60d919e)
> > #15 0x0000560286e35c64 llvm::SLPVectorizerPass::runImpl(llvm::Function&, llvm::ScalarEvolution*, llvm::TargetTransformInfo*, llvm::TargetLibraryInfo*, llvm::AAResults*, llvm::LoopInfo*, llvm::DominatorTree*, llvm::AssumptionCache*, llvm::DemandedBits*, llvm::OptimizationRemarkEmitter*) (build-all/bin/opt+0x60d6c64)
> > #16 0x0000560286e351e7 llvm::SLPVectorizerPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x60d61e7)
> > #17 0x000056028682b6cd llvm::detail::PassModel<llvm::Function, llvm::SLPVectorizerPass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilderPipelines.cpp:0:0
> > #18 0x00005602855c75c7 llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x48685c7)
> > #19 0x00005602868220fd llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilderPipelines.cpp:0:0
> > #20 0x00005602855cc19e llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x486d19e)
> > #21 0x00005602868187ed llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) PassBuilderPipelines.cpp:0:0
> > #22 0x00005602855c62b7 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x48672b7)
> > #23 0x00005602867a3b5c llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (build-all/bin/opt+0x5a44b5c)
> > #24 0x0000560285367c12 optMain (build-all/bin/opt+0x4608c12)
> > #25 0x00007f5db11387e5 __libc_start_main (/lib64/libc.so.6+0x3a7e5)
> > #26 0x000056028536582e _start (build-all/bin/opt+0x460682e)
> > Abort (core dumped)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > [bbi-104215.ll.gz](https://github.com/user-attachments/files/18839931/bbi-104215.ll.gz)
> 
> Must be fixed in [0e1ffa3](https://github.com/llvm/llvm-project/commit/0e1ffa397ef35fe55f7bd768d42f94241d1a2caa)

Yep, thanks.

**Pull request merged**